src/Frame.cc

```diff
@@ -1385,57 +1380,68 @@ void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save
+        if (save_file_msg.keep_degenerate()) {
+            keep_degenerate = true;
+        }
 
-        // Construct a new CASA image
         try {
-            auto out_image =
-                std::make_unique<casacore::PagedImage<casacore::Float>>(image->shape(), image->coordinates(), output_filename.string());
-            out_image->setMiscInfo(image->miscInfo());
-            out_image->setImageInfo(image->imageInfo());
-            out_image->appendLog(image->logger());
-            out_image->setUnits(image->units());
-            out_image->putSlice(temp_array, start);
-
-            // Copy the mask if the original image has
-            if (image->hasPixelMask()) {
-                casacore::Array<casacore::Bool> image_mask;
-                image->getMaskSlice(image_mask, slice);
-                out_image->makeMask("mask0", true, true);
-                casacore::Lattice<casacore::Bool>& out_image_mask = out_image->pixelMask();
-                out_image_mask.putSlice(image_mask, start);
+            if (region) {
+                auto latt_region_holder = LattRegionHolder(image_region);
+                auto slice_sub_image = GetExportRegionSlicer(save_file_msg, image_shape, region_shape, image_region, latt_region_holder);
+                _loader->GetSubImage(slice_sub_image, latt_region_holder, sub_image, keep_degenerate);
+            } else {
+                auto slice_sub_image = GetExportImageSlicer(save_file_msg, image_shape);
+                _loader->GetSubImage(slice_sub_image, sub_image, keep_degenerate);
             }
+            image = sub_image.cloneII();
         } catch (casacore::AipsError error) {
             message = error.getMesg();
             save_file_ack.set_success(false);
             save_file_ack.set_message(message);
             return;
         }
-        success = true;
-    } else if (output_file_type == CARTA::FileType::FITS) {
-        // Remove the old image file if it has a same file name
-        casacore::Bool ok = casacore::ImageFITSConverter::ImageToFITS(message, *image, output_filename.string(), 64, casacore::True,
-            casacore::True, -32, 1.0, -1.0, casacore::True, casacore::False, casacore::True, casacore::False, casacore::False,
-            casacore::False, casacore::String(), casacore::True);
-        if (ok) {
-            success = true;
-        }
     } else {
-        message = "No saving file action!";
+        return;
+    }
+
+    try {
+        std::unique_lock<std::mutex> ulock(_image_mutex); // Lock the image while saving the file
+        switch (output_file_type) {
+            case CARTA::FileType::CASA:
+                success = ExportCASAImage(*image, output_filename, message);
+                break;
+            case CARTA::FileType::FITS:
+                success = ExportFITSImage(*image, output_filename, message);
+                break;
+            default:
+                message = "Could not export file. Unknown file type";
```

Comment:
We should print out the file type here. Because it's an enum value, we probably need to add a function or  just an `unordered_map` to `Util` to map it to a string name. This may come in useful in other messages.

Reply:
Good idea. Will log a message "Export a CASA image" for example.

Reply:
I meant that you should log the type of file when the file *could not* be exported:

    "Could not export file. Unknown file type {}", ...

It's OK also to log it when exporting was successful.